===================================================================
--- wtk/src/org/apache/pivot/wtk/DesktopApplicationContext.java	(revision 1292513)
+++ wtk/src/org/apache/pivot/wtk/DesktopApplicationContext.java	(working copy)
@@ -273,7 +273,7 @@
     }
 
     // Native host dialog for secondary displays
-    private static class HostDialog extends java.awt.Dialog implements TitledWindow {
+    private static class HostDialog extends java.awt.Frame implements TitledWindow {
         private static final long serialVersionUID = 5340356674429280196L;
 
         private DisplayHost displayHost = new DesktopDisplayHost();
@@ -281,8 +281,8 @@
         private DisplayListener displayCloseListener;
 
         public HostDialog(java.awt.Window owner, boolean modal, DisplayListener displayCloseListener) {
-            super(owner, modal ?
-                java.awt.Dialog.ModalityType.APPLICATION_MODAL : java.awt.Dialog.ModalityType.MODELESS);
+            //super(owner, modal ?
+                //java.awt.Dialog.ModalityType.APPLICATION_MODAL : java.awt.Dialog.ModalityType.MODELESS);
 
             this.displayCloseListener = displayCloseListener;
 
